pkg/skaffold/docker/authenticators.go

```diff
@@ -0,0 +1,108 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package docker
+
+import (
+	"strings"
+	"sync"
+
+	"github.com/docker/cli/cli/config"
+	"github.com/google/go-containerregistry/pkg/authn"
+	"github.com/google/go-containerregistry/pkg/name"
+	"github.com/google/go-containerregistry/pkg/v1/google"
+)
+
+var authenticators = Authenticators{}
+
+// Authenticators stores an authenticator per registry.
+type Authenticators struct {
+	byRegistry map[string]*lockedAuthenticator
+	lock       sync.Mutex
+}
+
+// For retrieves the authentiator for a given image reference.
+func (a *Authenticators) For(ref name.Reference) authn.Authenticator {
+	registry := ref.Context().Registry.Name()
+
+	a.lock.Lock()
+	defer a.lock.Unlock()
+
+	// Get existing authenticator
+	if auth, present := a.byRegistry[registry]; present {
+		return auth
+	}
+
+	// Create a new authenticator
+	auth := &lockedAuthenticator{
+		delegate: a.newAuthenticator(ref),
+	}
+
+	if a.byRegistry == nil {
+		a.byRegistry = map[string]*lockedAuthenticator{}
+	}
+	a.byRegistry[registry] = auth
+
+	return auth
+}
+
+// lockedAuthenticator is an authn.Authenticator that can
+// be used safely from multiple go routines.
+type lockedAuthenticator struct {
+	delegate authn.Authenticator
+	lock     sync.Mutex
+}
+
+func (a *lockedAuthenticator) Authorization() (string, error) {
+	a.lock.Lock()
+	authorization, err := a.delegate.Authorization()
+	a.lock.Unlock()
+	return authorization, err
+}
+
+func (a *Authenticators) newAuthenticator(ref name.Reference) authn.Authenticator {
+	registry := ref.Context().Registry.Name()
+
+	// Search for the configured authenticator
+	cfg, err := config.Load(configDir)
+	if err == nil && (len(cfg.CredentialHelpers) > 0 || cfg.CredentialsStore != "" || len(cfg.AuthConfigs) > 0) {
+		// If the credential helper is `gcloud`, we use google.NewGcloudAuthenticator()
+		// It is more efficient because it reuses tokens.
+		if cfg.CredentialHelpers[registry] == "gcloud" {
+			if auth, err := google.NewGcloudAuthenticator(); err == nil {
+				return auth
+			}
+		}
+
+		// Call `Resolve` only when there's a non-empty `~/.docker/config.json`
+		// otherwise it prints annoying logs.
+		auth, err := authn.DefaultKeychain.Resolve(ref.Context().Registry)
+		if err == nil {
+			return auth
+		}
+	}
+
+	// No authenticator is configured:
+	// - check if we can use `gcloud`
+	// - otherwise, default to Anonymous
+	if registry == "gcr.io" || strings.HasSuffix(registry, ".gcr.io") {
```

Comment:
should we try and use gcloud before trying to use the default auth? that way we can also reuse the tokens if gcloud auth works, rather than using the less efficient default auth

```golang
if cfg.CredentialHelpers[registry] == "gcloud" || registry == "gcr.io" || strings.HasSuffix(registry, ".gcr.io")
```

Reply:
This is what previous lines tries to do

Reply:
go/container-registry doesn't make this easy

Reply:
yeah I meant just to combine them into one check. I would assume that most of the time, if the user is pushing to GCR then they have gcloud configured, so we may as well try that first.